api/src/main/java/au/org/aodn/nrmn/restapi/service/CorrectionService.java

```diff
@@ -0,0 +1,89 @@
+package au.org.aodn.nrmn.restapi.service;
+
+import au.org.aodn.nrmn.restapi.model.db.*;
+import au.org.aodn.nrmn.restapi.repository.ObservationRepository;
+import au.org.aodn.nrmn.restapi.repository.SurveyMethodRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Example;
+import org.springframework.stereotype.Service;
+
+import java.sql.Date;
+import java.sql.Time;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Service
+public class CorrectionService {
+    @Autowired
+    private ObservationRepository observationRepository;
+    @Autowired
+    private SurveyMethodRepository surveyMethodRepository;
+
+    public Stream<StagedRow> convertSurveyToStagedRows(Survey survey) {
+        List<SurveyMethod> surveyMethods = surveyMethodRepository.findAll(Example.of(
+                SurveyMethod.builder().survey(survey).build()));
+        return surveyMethods.stream().map(this::convertSurveyMethodToStagedRow);
+    }
+
+    private StagedRow convertSurveyMethodToStagedRow(SurveyMethod surveyMethod) {
+        Survey survey = surveyMethod.getSurvey();
+        StagedRow.StagedRowBuilder builder = StagedRow.builder()
+                .depth(getStagedDepth(survey.getDepth(), survey.getSurveyNum()))
+                .siteNo(survey.getSite().getSiteCode())
+                .siteName(survey.getSite().getSiteName())
+                .date(getDateString(survey.getSurveyDate()))
+                .time(getTimeString(survey.getSurveyTime()))
+                .vis(toString(survey.getVisibility()))
+                .direction(survey.getDirection())
+                .block(toString(surveyMethod.getBlockNum()))
+                .method(toString(surveyMethod.getMethod().getMethodId()))
+                .code(surveyMethod.getSurveyNotDone() ? "SND" : null);
+        List<Observation> observations = observationRepository.findAll(Example.of(
+                Observation.builder().surveyMethod(surveyMethod).build()));
+
+        observations.stream().filter(obs -> obs.getMeasure().getSeqNo() == 0)
+                .forEach(obs -> builder.inverts(obs.getMeasureValue().toString()));
+
+        observations.stream().findAny().ifPresent(obs -> builder
+                .species(getSpeciesScientificName(obs.getObservableItem()))
+                .diver(toString(obs.getDiver().getDiverId()))
+        );
+
+        builder.measureJson(observations.stream()
+                .filter(obs -> obs.getMeasure().getSeqNo() == 0)
+                .collect(Collectors.toMap(
+                        obs -> obs.getMeasure().getSeqNo(),
+                        obs -> obs.getMeasureValue().toString())));
+
+        return builder.build();
+    }
+
+
+    private String toString(Object object) {
+        return object == null ? null : object.toString();
+    }
+
```

Comment:
this method can be removed and use instead String.valueOf(object)

Reply:
Sadly String.valueOf(object) throws a null pointer exception when given a null

Reply:
Whoops, I mean not a null pointer, but the string "null" which is still not exactly what I need.